### ISSUE TITLE
Is SSO integration available?

### ISSUE THREAD

**Issue by audetto** [cronicle-edge/cronicle-edge#41]
I mean something where I can delegate to a company service the user authentication.

1) is token present in cookies?
2) if not redirect to sso url (and go back to 1)
3) check with external system if token is valid
4) if yes, user is authenticated

**Comment by mikeTWC1984** [comment]
No, neither original cronicle or this fork provide SSO.  I added a simple LDAP/AD for external authentication though.  Are you using some cloud auth provider?

**Comment by audetto** [comment]
Ok.
No, on my organisation I can delegate all of this to an internal system.

If the mainpage redirects to

``http://internal.sso.provider/?returnUrl=http://address.of.cronicle``

then, when it comes back to `http://address.of.cronicle` there will be a cookie with a magic token.
At this point cronicle could verify the token and it is all done without anyone typing any password.

If this is too complicated, would it be possible to outsource to a new js module (that I can write for myself) the verification of the password?
Rather than cronicle having its own set of passwords, I could implement a function that checks if the password is correct, again delegating it to the company internal authentication provider.

**Comment by mikeTWC1984** [comment]
I see. I think SSO is not that trivial to implement (in general). Might be an overkill for such indie tool. But if I get a chance, I'll definitely look into this.

If you just want to do custom password checking, you can take a look at my user.js implementation. It's pretty much dual mode, if you indicate user as external auth would go in custom route:
https://github.com/cronicle-edge/cronicle-edge/blob/main/lib/user.js#L210
You just need to replace do_ad_auth with your custom code. I also have some internal plumbing for external logins. Just keep in mind - you still need to register user and his permissions within cronicle  (check external box when creating user)

**Comment by audetto** [comment]
>I see. I think SSO is not that trivial to implement (in general).

I would like to disagree with this statement: cronicle must already be using cookies to decide if a user is autheticated or it should be redirected to the (cronicle) login page.

If I could inject a function in that process which is allowed to inspect the coockies, I could check if the company SSO cookie is present (and valid) and in that case return the (trusted) name of the user (which of course must be known to cronicle).

If handling the redirection on a failure is too complicated, then it could continue as normal to the password page.

I would then bookmark a different URL for Cronicle, not `http://cronicle.company.com` but `http://sso.service/returUrl=http://cronicle.company.com` which means that when the code gets to the above function the token is always there.

https://github.com/cronicle-edge/cronicle-edge/blob/778b47c7a6187562d7c7e19b49654577aadc4d96/lib/user.js#L221

it this function were given the coockies, I could do what I need, maybe not as elegant as possible, but working.
I would ignore the password, check the token is valid and the username match.

Is there an example how to override that function with a configuration without touching the source code?

**Comment by mikeTWC1984** [comment]
Cronicle is using LocalStorage object, where it puts session_id property for auth. So it's not using cookies, although it should be hard to tweak it to switch to cookies.
Regarding login page redirect - it's not a true redirect, it just loading hidden login tab:
https://github.com/cronicle-edge/cronicle-edge/blob/778b47c7a6187562d7c7e19b49654577aadc4d96/htdocs/js/pages/Base.class.js#L44

I don't have much experience with implementing  SSO , but as far as I know you still should register you app with SSO provider (get client id/key), then do some signature generation on your app end before you do that redirection. So it's obviously not something impossible, but not that easy as password checking. 

>Is there an example how to override that function with a configuration without touching the source code?
 You can create your custom version of User.js if you want, and change path to it in lib/main.js file.

**Comment by audetto** [comment]
I did not realise cronicle was using LocalStorage.

>I don't have much experience with implementing SSO , but as far as I know you still should register you app with SSO provider (get client id/key), then do some signature generation on your app end before you do that redirection. So it's obviously not something impossible, but not that easy as password checking.

I am sure there are plenty of different authorisation systems, and I don't know them enough to comment on how general ours is.

We do not need to generate any key, nor signatures.
If the token is present in the cookies, then I just have to check it is legit (via a simple webservice call) and I know exactly who is trying to log in.

For more details, I can tell you than the single thing I need to do to make this work is to register the return url, as otherwise the sso service will refuse to redirect.

`http://sso.service/returUrl=http://cronicle.company.com`

The url `http://cronicle.company.com` must be registered, but it is nothing cronicle needs to know about.

No password is ever entered as it is all managed via the handshake between the user browser and the sso service (all transparent from the cronicle point of view). This does not work from a VM because the handshake does not work, but it is irrelevant as all cases we care about come from users logging in from a real corporate machine.

**Comment by spali** [comment]
I would love something similar.
Maybe a more general approach would be more flexible with less code.
My usecase would be to use the proxy in front of cronicle to do the authentication and forward the user and optionally groups in the header.

On application side, you usually configure ip's or cidr ranges for trusted proxy addresses. Without these matching against the connected IP, you just ignore the auth headers.
Additionally configure the header names containing at least the username (something like `X-Remote-User`).
for groups this would be possible by the header name (i.e. `X-Remote-Group`) and a delimiter for the groups (i.e. `X-Remote-Group-Delim` and content  `,` or `|`.
You can extend this and make it as configurable as you want 😉 

Implementation would need to check ip of the connection against the trusted ip/cidr list and read the headers. If headers exists and everything is there we can just open a session for the provided user... no login screen nothing.
Depending on the API between frontend and backend it may need client side code forward every request to the backend where the backend can do this.

In my eyes, this approach is more flexible than implementing any possible SSO solution or standard of tokens etc.... Almost every standard can be somehow used by a reverse proxy in front of the application and provide the user (and maybe more) standardized back to the application.
i.e. nginx supports this as well as traefik etc. With differrent auth providers like authentik, authelia, keycloak etc.

This functionality could also be provided by plugins if on client and on server side everything is there. 
If on server side, every request (authenticated or not) passes a plugin function which has access to the headers and connection ip (usually the http request object) then the plugin can do this.

so instead of an auth plugin like `function(user,password) { return true/false }`, it would need `function(httpRequest, headerconfig) {return true/false}` at least on any request that is not already authenticated. After auth, it may reely on the default implementation of cronicle already. But some implementations check this on every request that comes in, to also allow a user to be logged out.

I assume audetto's SSO solution should also work with this if combined with a reverse proxy.

**Comment by mikeTWC1984** [comment]
@spali  @audetto 
I was working on oath2 SSO, it's pretty much ready, will release it soon (maybe this week)

@spali  - the approach you are suggesting should be simple to integrate into that new oauth workflow. The only thing - using trusted IP for "authorization" doesn't sound like a greatest idea. Which auth tool you are using? As far as I can see authelia can add jwt token to the header as well (like you can use some shared secret between cronicle and authelia to generate/verify token)
If you using authentic it can be used directly for oauth (without reverse proxy), actually this is what I used.

**Comment by spali** [comment]
I use athelia in combination with traefik forwardAuth.
Also started to use authentic with forward auth.
I like forward auth because of its simplicity which helps to prevent handling errors.

**Comment by audetto** [comment]
> so instead of an auth plugin like `function(user,password) { return true/false }`, it would need `function(httpRequest, headerconfig) {return true/false}` at least on any request that is not already authenticated. After auth, it may reely on the default implementation of cronicle already. But some implementations check this on every request that comes in, to also allow a user to be logged out.
> 
> I assume audetto's SSO solution should also work with this if combined with a reverse proxy.

With 2 caveats:

1. we need to agree which `users` we are authenticating
2. what happens if the function returns `false`? password request? maybe allow some redirect

So a function like

```
function(user, httpRequest, headerconfig) {
    return true/false/url
}
```

or, if you don't want the `user` argument, then the return must be: `return user/false/url`

**Comment by spali** [comment]
@audetto 
> 2. what happens if the function returns `false`? password request? maybe allow some redirect

depends on how flexible it should be, easiest would be to return false and cronicle calls the next plugin or falls back to it's own login.

> ```
> function(user, httpRequest, headerconfig) {
>     return true/false/url
> }
> ```
the app (cronicle) does not require an url to forward to. the proxy does forward to the auth providers login and gets redirected back to the origin url but this time (if authenticated) with the additional headers sent from the proxy to the app.

----------


@mikeTWC1984 Nope, `document.cookie` is an empty string. But it was expected, as the forward auth provides (authelia, authentik etc.) can be a complete different domain, so no cookie to check. It would be possible in some scenarios were you could see the cookie in the header sent by the browser, depending on the cookie domain of the auth provider and cronicle, but that's not guaranteed.  The only information exchanged to the app (cronicle) are the headers in every request. 
In my setup cronicle would just get the following example headers from the reverse proxy  (info is from authelia):
```
Remote-Email: admin@example.com
Remote-Groups: admins,users
Remote-Name: Admin User
Remote-User: admin
```

Authentik gives a bit more (also more complex options, has a lot more features, but the relevant for forward auth are thee same except different header names and different groups delimiter):
```
X-Authentik-Email: admin@example.com
X-Authentik-Groups: admin|users
X-Authentik-Name: Admin User
X-Authentik-Uid: <long number>
X-Authentik-Username: admin
```

Minimal implementation would have to get the username and check if internally available, if yes create session for this user.
1. optional, if user does not exists create it. Optional get additional details for the user from the headers (like email, name etc).
2. optional, sync this additional details to the internal user everytime the session is new.
3. optional read the groups and sync them also to the internal user.
If user or groups pre-exists, or is created on the fly when first seen, is a decision for the implementation.

And not to forget, only accept these headers if the IP is trusted as configured. As for any reverse proxy, the real client id is usually sent by the proxy also in the headers if required for logging the real clients ip address and not only the proxy address.

**Comment by audetto** [comment]
Yes, such a function https://github.com/cronicle-edge/cronicle-edge/issues/41#issuecomment-2489636218 would work.
And I can do my own things, like checking some IP range, if some cookie is good etc...

In a previous reply https://github.com/cronicle-edge/cronicle-edge/issues/41#issuecomment-2489554305

>the app (cronicle) does not require an url to forward to. the proxy does forward to the auth providers login and gets redirected back to the origin url but this time (if authenticated) with the additional headers sent from the proxy to the app.

yes yes, totally agree.

My url was simply to send failures to some "instruction page" explaining how to get access.

Because in my case, a failure to login cannot be solved with a password: there are no passwords.
So I wanted to avoid people coming with "What's my password?", when the solution (in my case) is to integrate with whatever we have behind the proxy, which Cronicle cannot solve.

So the page could simply say: "Raise a new internal ticket of type XYZ, and ask permission to John Smith".
But we can live without this.

**Comment by spali** [comment]
> My url was simply to send failures to some "instruction page" explaining how to get access: "e.g. Raise ticket of type XYZ". But this can all be sorted out in the proxy layer too, so we can live without it.

If @mikeTWC1984 agrees to an api including a modifiable response object, similar to expressjs middleware, then this and much more is possible.
expressjs middleware can alter the response (including redirect or editing the response body) and give the control back by calling `next()` (next registered middleware in the list get's control) or end the request by calling `req.end()`. Or pass an error to the `next(new Error())`.
So the fundamental api could be same as a expressjs middleware, just somehow with a defined handover of a user info object. i.e. `next(userObj)`.
Also it should be defined if possible to have multiple such middlewares and if cronicle does call all in the chain under which conditions.
BTW: instead of the response object, it could be also a abstract version of it... to not give the middleware to much control... but would require more implementation...

**Comment by mikeTWC1984** [comment]
After some research, I realize authelia is adding session cookie, it just not accessible from js, but it will be attached in the following request. 

Now, the problem is that cronicle is single page app (just a piece of front end js, not a backend), so it cannot handle header info

I guess best approach here is to redirect request to some handler (not to the original url)
E.g. if user requests  https://cron.example.com/#Schedule
it should be redirected to uri like https://cron.example.com/auth/authelia  first
I guess info about original path will be stored in headers, so it can be redirected there after processing.

@spali I guess that should be doable, can you test?

**Comment by audetto** [comment]
This is exactly the same as my use case.

1. the final url is encoded in a query parameter to the proxy (e.g. `?redirectUrl=http://address.of.cronicle`)
2. proxy does its magic (or fails of course)
3. redirects to the url from 1
4. puts authentication data (username, etc) in the header (for me in the cookie section, but I guess it could vary)
5. cronicle checks data from 4 is valid

Cronicle is already using a similar approach at point 5 to avoid multiple password requests for users who have already authenticated, but it is using `LocalStorage`, not http header. Fine, different locations, but conceptually the same.

"If I find some magic data already in the "page", I skip the password request."

Whether it is `LocalStorage` or http header (or cookies), could be abstracted away so that via a plugin one can inject custom code.

**Comment by mikeTWC1984** [comment]
OK, after some experiments, I think this could be achieved. As I mentioned, by default when you go to "/" route in cronicle, it's basically just serving static content from htdocs folder. In other words cronicle web UI is just index.html page, which cannot check headers. 
The webserver behind cronicle (pixl-server-web) doesn't seem to have any middleware options, especially for static content. However, looks like you can redefine "/" as separate route in this webserver, just to load index.html and check headers before that, which can be checked and exchanged to cronicle session. 
The only problem now is that autelia session and cronicle session need to be synchronize somehow. Perhaps same cookie could be used for both authelia and cronicle. 
Anyway, I'll share some code snippet later you can add to user.js to play around

**Comment by mikeTWC1984** [comment]
Yes, that's the problem - there is no hook/middleware to check headers/cookies when user goes to / route, at least I cannot see any good option in docs:
https://github.com/jhuckaby/pixl-server-web?tab=readme-ov-file#custom-uri-handlers

In order to load user profile user should have session_id in localStorage or cookie. Normally user gets it by clicking login button. If you want to load a profile without user interaction session cookie should be attached to response from "/".  

The way you suggest to handle that (as it sounds to me) is basically making cronicle authless,  just define  user context via headers. I guess the easiest way to achieve it is to exchange header info to session, so same internal mechanisms will wok seamlessly (that could be achieved with a hack i suggested with registering / as route). To be honest it's the only way I can imagine at this point.

**Comment by spali** [comment]
Ok... checked the requests what cronicle does from the client side. Do know understand a bit more what you try to tell me.
I never have seen it via local storage and sent by POST requests and errors come back by simple json without proper http status code. But shouldn't make a difference for us in this case here.
The sent session_id in the payload has to be checked somewhere in the backend on every request. Because if I call any /api endpoint directly (no payload with session id is sent) it just sends me back a response json object which represents an error for the missing session.
This decision in the backend has to be somewhere and we need to find then a way to hook into this decision.

**Comment by spali** [comment]
I think found the code:
https://github.com/cronicle-edge/cronicle-edge/blob/5dc439d3448bf2033e8c6c16c3c976336e21c914/lib/api.js#L498

Looks like this function has access to the request and returns a session and a user object by callback at the end... 
So we need a hook function that is called there somewhere there.
At least before line 576 or even before the other checks above?
So we need to define (or find a way) to getting from the hook function which returns at least a username a session and user object to pass to the callback like in line 594.

**Comment by rkvasi** [comment]
CAS support is avaialble in this fork https://github.com/rkvasi/cronicle-edge

I will raise a pull request soon

**Comment by rkvasi** [comment]
> I think found the code:
> 
> [cronicle-edge/lib/api.js](https://github.com/cronicle-edge/cronicle-edge/blob/5dc439d3448bf2033e8c6c16c3c976336e21c914/lib/api.js#L498)
> 
> Line 498 in [5dc439d](/cronicle-edge/cronicle-edge/commit/5dc439d3448bf2033e8c6c16c3c976336e21c914)
> 
>  loadSession: function (args, callback) { 
> Looks like this function has access to the request and returns a session and a user object by callback at the end... So we need a hook function that is called there somewhere there. At least before
> 
> [cronicle-edge/lib/api.js](https://github.com/cronicle-edge/cronicle-edge/blob/5dc439d3448bf2033e8c6c16c3c976336e21c914/lib/api.js#L576)
> 
> Line 576 in [5dc439d](/cronicle-edge/cronicle-edge/commit/5dc439d3448bf2033e8c6c16c3c976336e21c914)
> 
>  // finally check for API key and return error if no session/api key detected 
> or even before the other checks above?
> So we need to define (or find a way) to getting from the hook function which returns at least a username a session and user object to pass to the callback like in
> [cronicle-edge/lib/api.js](https://github.com/cronicle-edge/cronicle-edge/blob/5dc439d3448bf2033e8c6c16c3c976336e21c914/lib/api.js#L594)
> 
> Line 594 in [5dc439d](/cronicle-edge/cronicle-edge/commit/5dc439d3448bf2033e8c6c16c3c976336e21c914)
> 
>  callback(null, session, user);

https://github.com/cronicle-edge/cronicle-edge/issues/41#issuecomment-3024000506
